libsolidity/codegen/ir/IRLValue.cpp

```diff
@@ -166,3 +180,79 @@ string IRStorageArrayLength::setToZero() const
 {
 	return storeValue("0", *TypeProvider::uint256());
 }
+
+IRMemoryItem::IRMemoryItem(
+	IRGenerationContext& _context,
+	std::string _address,
+	bool _calldata,
+	bool _byteArrayElement,
+	Type const& _type
+):
+	IRLValue(_context, &_type),
+	m_address(move(_address)),
+	m_byteArrayElement(_byteArrayElement),
+	m_calldata(_calldata)
+
+{ }
+
+string IRMemoryItem::retrieveValue() const
+{
+	solUnimplementedAssert(m_byteArrayElement, "Byte arrays not yet implemented!");
+
+	if (m_type->isValueType())
+		return m_context.utils().readFromMemory(*m_type) +
+			"(" +
+			m_address +
+			")\n";
+	else if (m_calldata)
+		return "calldataload(" + m_address + ")\n";
+	else
+		return "mload(" + m_address + ")\n";
+}
+
+string IRMemoryItem::storeValue(string const& _value, Type const& _type) const
+{
+	solAssert(!m_calldata, "Can't store value in calldata memory!");
+
+	if (!m_type->isValueType())
+	{
+		solUnimplementedAssert(_type == *m_type, "Conversion not implemented for assignment to memory.");
+
+		solAssert(m_type->sizeOnStack() == 1, "");
+		return "mstore(" + m_address + ", " + _value + ")\n";
+	}
+
+	solAssert(_type.isValueType(), "");
+
+	string converted = _value;
+
+	solAssert(_type == *m_type, "");
+
+	if (_type != *m_type)
+		converted =
+			m_context.utils().conversionFunction(_type, *m_type) +
+			"(" +
+			_value +
+			")";
+
```

Comment:
Maybe we should add cleanup. Did we talk about that before?

Reply:
We did, but in a different context. I'll add it